packages/SwingSet/src/kernel/vatManager/supervisor-subprocess-xsnap.js

```diff
@@ -0,0 +1,226 @@
+// @ts-check
+import { assert, details } from '@agoric/assert';
+import { importBundle } from '@agoric/import-bundle';
+import { Remotable, getInterfaceOf, makeMarshal } from '@agoric/marshal';
+// grumble... waitUntilQuiescent is exported and closes over ambient authority
+import { waitUntilQuiescent } from '../../waitUntilQuiescent';
+
+import { makeLiveSlots } from '../liveSlots';
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+// eslint-disable-next-line no-unused-vars
+function workerLog(first, ...args) {
+  // @ts-ignore
+  // eslint-disable-next-line
+  // print(`---worker: ${first}`, ...args);
+}
+
+workerLog(`supervisor started`);
+
+/**
+ * @param { (cmd: ArrayBuffer) => ArrayBuffer } issueCommand
+ * @typedef { [unknown, ...unknown[]] } Tagged
+ * @typedef { (item: Tagged) => unknown } DataHandler
+ * @typedef { (item: Tagged) => Promise<Tagged> } AsyncHandler
+ */
+function ManagerPort(issueCommand) {
+  /** @type { (item: Tagged) => ArrayBuffer } */
+  const encode = item => encoder.encode(JSON.stringify(item)).buffer;
+
+  /** @type { (msg: ArrayBuffer) => any } */
+  const decodeData = msg => JSON.parse(decoder.decode(msg));
+
+  /** @type { (msg: ArrayBuffer) => Tagged } */
+  function decode(msg) {
+    /** @type { Tagged } */
+    const item = decodeData(msg);
+    assert(Array.isArray(item), details`expected array`);
+    assert(item.length > 0, details`empty array lacks tag`);
+    return item;
+  }
+
+  return harden({
+    /** @type { (item: Tagged) => void } */
+    send: item => {
+      issueCommand(encode(item));
+    },
+    /** @type { DataHandler } */
+    call: item => decodeData(issueCommand(encode(item))),
+    /** @type { (f: AsyncHandler)  => ((msg: ArrayBuffer) => { result?: ArrayBuffer })} */
+    handler: f => msg => {
+      const report = {};
+      f(decode(msg))
+        .then(item => {
+          workerLog('result', item);
+          report.result = encode(item);
+        })
+        .catch(err => {
+          report.result = encode(['err', err.message]);
```

Comment:
I'm hesitant to mess with the success case; that's the one that has all the tests working and it would involve changes all over the place. A rejected promise here is a programming error, AFAIK. Maybe it should just halt and catch fire. But I was looking at it as: `item` is already of type `['ok', T] | ['err', string]`, so another `err` case would still fit.